clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp

```diff
@@ -2525,25 +2565,49 @@ RValue CIRGenFunction::emitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
 
   // See if we have a target specific intrinsic.
   std::string Name = getContext().BuiltinInfo.getName(BuiltinID);
-  Intrinsic::ID IntrinsicID = Intrinsic::not_intrinsic;
+  Intrinsic::ID intrinsicID = Intrinsic::not_intrinsic;
   StringRef Prefix =
       llvm::Triple::getArchTypePrefix(getTarget().getTriple().getArch());
   if (!Prefix.empty()) {
-    IntrinsicID = Intrinsic::getIntrinsicForClangBuiltin(Prefix.data(), Name);
+    intrinsicID = Intrinsic::getIntrinsicForClangBuiltin(Prefix.data(), Name);
     // NOTE we don't need to perform a compatibility flag check here since the
     // intrinsics are declared in Builtins*.def via LANGBUILTIN which filter the
     // MS builtins via ALL_MS_LANGUAGES and are filtered earlier.
-    if (IntrinsicID == Intrinsic::not_intrinsic)
-      IntrinsicID = Intrinsic::getIntrinsicForMSBuiltin(Prefix.data(), Name);
+    if (intrinsicID == Intrinsic::not_intrinsic)
+      intrinsicID = Intrinsic::getIntrinsicForMSBuiltin(Prefix.data(), Name);
   }
 
-  if (IntrinsicID != Intrinsic::not_intrinsic) {
+  if (intrinsicID != Intrinsic::not_intrinsic) {
     unsigned iceArguments = 0;
     ASTContext::GetBuiltinTypeError error;
     getContext().GetBuiltinType(BuiltinID, error, &iceArguments);
     assert(error == ASTContext::GE_None && "Should not codegen an error");
-    if (iceArguments > 0)
+
+    llvm::StringRef name = llvm::Intrinsic::getName(intrinsicID);
+    // cir::LLVMIntrinsicCallOp expects intrinsic name to not have prefix
+    // "llvm." For example, `llvm.nvvm.barrier0` should be passed as
+    // `nvvm.barrier0`.
+    if (!name.consume_front("llvm."))
+      assert(false && "bad intrinsic name!");
+
+    cir::FuncType intrinsicType =
+        getIntrinsicType(&getMLIRContext(), intrinsicID);
+
+    SmallVector<mlir::Value> args;
+    for (unsigned i = 0; i < E->getNumArgs(); i++) {
+      mlir::Value arg = emitScalarExpr(E->getArg(i));
+      if (arg.getType() != intrinsicType.getInput(i))
```

Comment:
Can you try to match OG a bit more? As you noticed I could figure this out in previous iterations:
```
mlir::Type pTy = ... 
      if (pTy != ArgValue->getType()) {
```